compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1039,9 +1058,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                       (tycon1sym.isClass || tycon2sym.isClass)
                       && (!touchedGADTs || gadtIsInstantiated)
 
-                    inFrozenGadtIf(!tyconIsInjective) {
-                      isSubArgs(args1, args2, tp1, tparams)
-                    }
+                    def checkArgs(): Boolean =
+                      if tycon1sym == tycon2sym && tycon1sym.isAliasType then
+                        // if types are simple forwarding aliases then try the dealiased types
+                        tp1.superType match
+                          case tp1a as AppliedType(_, args1a) if args1a.eqElements(args1) =>
+                            tp2.superType match
+                              case tp2a as AppliedType(_, args2a) if args2a.eqElements(args2) =>
+                                return recur(tp1a, tp2a)
+                              case _ =>
+                          case _ =>
+                        // Otherwise either compare arguments or compare dealiased types.
+                        // Note: This test should be done with an `either`, but this fails
+                        // for hk-alias-unification.scala The problem is not GADT boundschecking;
+                        // that is taken care of by the outer inFrozenGadtIf test. The problem is
+                        // that hk-alias-unification.scala relies on the right isSubArgs being performed
+                        // when constraining the result type of a method. But there we are only allowed
+                        // to use a necessaryEither, since it might be that an implicit conversion will
```

Comment:
You are right. The change was actually introduced here: https://github.com/lampepfl/dotty/pull/8635

It was a way to avoid overcommitting when constraining the result type. Unfortunately, it seems that this causes the problems we encounter here. So, what to do? All this is getting more and more ad-hoc, which is not very satisfactory.